silkworm/db/kv/txn_num.cpp

```diff
@@ -35,12 +35,19 @@ static Task<TxNum> last_tx_num_for_block(const std::shared_ptr<kv::api::Cursor>&
                                          BlockNum block_number,
                                          chain::CanonicalBodyForStorageProvider canonical_body_for_storage_provider) {
     const auto block_number_key = block_key(block_number);
-    const auto key_value = co_await max_tx_num_cursor->seek_exact(block_number_key);
+    auto key_value = co_await max_tx_num_cursor->seek_exact(block_number_key);
     if (key_value.value.empty()) {
         SILKWORM_ASSERT(canonical_body_for_storage_provider);
         auto block_body_data = co_await canonical_body_for_storage_provider(block_number);
         if (!block_body_data) {
-            throw std::invalid_argument("Bad block_num  " + std::to_string(block_number));
+            key_value = co_await max_tx_num_cursor->last();
+            if (key_value.value.empty()) {
+                co_return 0;
+            }
+            if (key_value.value.size() != sizeof(TxNum)) {
+                throw std::length_error("Bad TxNum value size " + std::to_string(key_value.value.size()) + " in db");
+            }
+            co_return endian::load_big_u64(key_value.value.data());
```

Comment:
Returning the max transaction number associated to the last block when the target block cannot be found neither in db nor in snapshots is OK for `min_tx_num` and `max_tx_num`, but not for `block_num_from_tx_num`.
I would suggest to change the return type to `Task<std::optional<TxNum>>` and use `std::nullopt` here, relying on `last` call outside when needed.